# Learning From Data - Homework 5
## Ognen Nastov

<img src='hw5_images/hw5p1.png'>

**Answer:**

Substituting: 

$$0.008 < 0.1^2\left(1-\frac{9}{N}\right)$$

$$N > \frac{9}{1-\frac{0.008}{0.1^2}} = 45$$

Answer is 100, [c].

---

Extra: derive the equation (book exercise 3.4).

a) Show that in sample estimate of $y$ is $X w^{*} + H \epsilon$.

In sample estimate $\hat{y}$ is:

$$\begin{aligned}
\hat{y} & = Hy = H(w^{*T}x + \epsilon) \\
& = H w^{*T}x + H\epsilon =  Xw^{*} + H \epsilon
\end{aligned}$$

since $H w^{*T}x = H X w^{*} = X (X^T X)^{-1} (X^T X) w^{*} = X w^{*}$

b) Show that in-sample error $\hat{y}-y$ can be expressed as matrix times $\epsilon$.

$$\begin{aligned}
\hat{y}-y = Hy-y & = X w^* + H E - (X w^* + E) \\
& = (H-I)E
\end{aligned}$$

c) Express $E_{in}(w_{lin})$ in terms of $\epsilon$.

$$\begin{aligned}
E_{in}(w_{lin}) & = \frac{1}{N} {||\hat{y} - y||}^2 \\
& = \frac{1}{N}(\hat{y}-y)^T (\hat{y}-y) \\
& = \frac{1}{N} E^T (H-I)^T (H-I) E \\
& = \frac{1}{N} E^T (I-H) E \\
& = \frac{1}{N} E^T E - \frac{1}{N} E^T H E
\end{aligned}$$

since $H$ is symmetric, i.e. $H^T=H$, and also since $(I-H)^K = (I-H)$

d) Prove that $\mathbb E_D[E_{in}(w_{lin})] = \sigma^2 \left(1-\frac{d+1}{N}\right)$ using (c) and the independence of $\epsilon_1,\cdots,\epsilon_N$. Hint: use $\text{trace}(H) = d+1$.

$$\begin{aligned}
\mathbb E_D[E_{in}(w_{lin})] & = \mathbb E_D\left[\frac{1}{N}E^T E\right] - \mathbb E_D\left[\frac{1}{N}E^T H E\right] \\
& = \frac{1}{N} N \sigma^2 - \mathbb E_D\left[\frac{1}{N} E^T H E\right] \\
& = \frac{1}{N} N \sigma^2 - \frac{1}{N} \mathbb E_D\left[\sum_{i=1}^N H_{ii} \epsilon_i^2 + \sum_{i,j}^{i≠j} H_{ij} \epsilon_i \epsilon_j \right] \\
& = \frac{1}{N} N \sigma^2 - \frac{1}{N}\text{trace}(H) \sigma^2 \\ 
& = \sigma^2 \left(1 - \frac{d+1}{N}\right)
\end{aligned}$$

since $\mathbb E_D[\sum_{i,j}^{i≠j} H_{ij}\epsilon_i \epsilon_j] = 0$.

---

<img src='hw5_images/hw5p2a.png'>
<img src='hw5_images/hw5p2b.png'>

**Answer:**

$$g(x) = \text{sign}(\tilde{w}^T \Phi(x)) = \text{sign}(\tilde{w}_0 + \tilde{w}_1 x_1^2 + \tilde{w}_2 x_2^2)$$

For $x_1=0$, $g(x)$ is always $+1$. Which means $\tilde{w}_2 > 0$.  
For $x_2=0$, $g(x)$ is $+1$ at $(0,0)$, and negative for large positive or negative $x_1$. Which means $\tilde{w}_1 < 0$.

Answer is [d].

----

<img src='hw5_images/hw5p3.png'>

**Answer:**

$d+1 = 3$, thus $d_{vc} = d+1 = 3$

$\tilde{d}+1 = 15$, thus $\tilde{d}_{vc} ≤ \tilde{d}+1 = 15$

Answer is 15, i.e. [c].

----

<img src='hw5_images/hw5p4.png'>

**Answer:**

$$\frac{\partial E}{\partial u} = 2(u e^v - 2 v e^{-u}) (e^v + 2 v e^{-u})$$

Answer is [e].

----

<img src='hw5_images/hw5p5a.png'>
<img src='hw5_images/hw5p5b.png'>

**Answer:**

$$\frac{\partial E}{\partial v} = 2 (u e^v - 2 v e^{-u}) (u e^v - 2 e^{-u})$$

In [1]:
import numpy as np
import time

# problems 5 & 6
def E(u,v):
    E = (u*np.exp(v) - 2*v*np.exp(-u))**2
    return E;

def dEdu(u,v):
    value = 2*(u*np.exp(v) - 2*v*np.exp(-u))*(np.exp(v) + 2*v*np.exp(-u))
    return value

def dEdv(u,v):
    value = 2*(u*np.exp(v) - 2*v*np.exp(-u))*(u*np.exp(v) - 2*np.exp(-u))
    return value

# gradient descent of E(u,v)
def grad_descent(E, dEdu, dEdv, uv_init, max_iters, learning_rate, max_error):
    iter = 0;
    u = uv_init[0]
    v = uv_init[1]
    while iter < max_iters:
        iter+=1
        delta_u = -dEdu(u,v)
        delta_v = -dEdv(u,v)
        u+= learning_rate*delta_u
        v+= learning_rate*delta_v
        if E(u,v) < max_error:
            break
    print(f"# iters = {iter}, error = {E(u,v)}, (u,v) = {(u,v)}")

In [2]:
grad_descent(E, dEdu, dEdv, (1,1), 20, 0.1, 1e-14)

# iters = 10, error = 1.2086833944220747e-15, (u,v) = (0.04473629039778207, 0.023958714099141746)


Answer is 10, i.e. [d].

---

<img src='hw5_images/hw5p6.png'>

In [3]:
grad_descent(E, dEdu, dEdv, (1,1), 20, 0.1, 1e-14)

# iters = 10, error = 1.2086833944220747e-15, (u,v) = (0.04473629039778207, 0.023958714099141746)


Answer is [e].

----

<img src='hw5_images/hw5p7.png'>

**Answer:**

In [4]:
def coord_descent(E, dEdu, dEdv, uv_init, max_iters, learning_rate, max_error):
    iter = 0;
    u = uv_init[0]
    v = uv_init[1]
    while iter < max_iters:
        iter+=1
        # step 1
        delta_u = -dEdu(u,v)
        u+= learning_rate*delta_u
        # step 2
        delta_v = -dEdv(u,v)
        v+= learning_rate*delta_v
        if E(u,v) < max_error:
            break
    print(f"# iters = {iter}, error = {E(u,v)}, (u,v) = {(u,v)}")    

In [5]:
coord_descent(E, dEdu, dEdv, (1,1), 15, 0.1, 1e-14)

# iters = 15, error = 0.13981379199615315, (u,v) = (6.29707589930517, -2.852306954077811)


Answer is 1e-1, [a].

----

<img src='hw5_images/hw5p8a.png'>
<img src='hw5_images/hw5p8b.png'>

**Answer:**

Using `N_out = 2000`, `max_epochs = 10000`.

In [6]:
# problem 8 & 9 - logistic regression with stochastic gradient descent

# dimension of input space d
# number of data points = N
# X is shape [N, d+1] i.e. [N, 3]
def make_data_set(N, d):
   X = np.zeros([N,d])
   X[:,0] = np.ones([N,])
   for i in range(1,d):
       Xi = np.random.uniform(-1,1,N)
       X[:,i] = Xi
   return (X)
    
# the target function
# d = 2
# sb is shape [2]
def make_line():
    x_1,y_1 = np.random.uniform(-1,1,2)
    x_2,y_2 = np.random.uniform(-1,1,2)
    slope = (y_1 - y_2) / (x_1 - x_2)
    b = y_2 - slope * x_2
    return (slope, b)

# y is shape [N]    
def classify_X(line, X):
    slope, b = line
    y = np.sign(X[:,1]*slope + b - X[:,2])
    return(y)
  
# print flags
print_lg = False

In [7]:
def logistic_regression(X, line, y, w_init, max_epochs, eta, max_error, N_out):
    N = np.size(X,0)
    d = np.size(X,1) - 1
    indices = np.arange(0,N)
    w = w_init
    n_epochs = 0
    while (1):
        n_epochs += 1
        w_old = w
        # permute the sample order
        indices_perm = np.random.permutation(indices)
        for index in indices_perm:
            grad_e_n = -y[index]*X[index] / \
                (1 + np.exp(y[index]*np.dot(w,X[index])))
            w = w - eta * grad_e_n
        delta_w_norm = np.sqrt(np.sum((w - w_old)**2))
        if delta_w_norm < max_error:
            break
    # generate out of sample
    X_out = make_data_set(N_out, d+1)
    y_out = classify_X(line, X_out)
    # estimate E_out
    E_out = (1/N_out)*np.sum(np.log(1 + np.exp(-y_out*np.dot(X_out,w))))
    if (print_lg):
        print(f"Epochs = {n_epochs} , E_out = {E_out}")
    return (n_epochs, E_out)

In [8]:
# N = 100, d = 2
def run_exp(N, d, num_runs):
    time_start = time.time()
    n_epochs_runs = np.empty(0)
    E_out_runs = np.empty(0)
    while (num_runs > 0):
        num_runs-= 1
        # generate target
        X = make_data_set(N, d+1)
        line = make_line()
        y = classify_X(line, X)
        w_init = np.zeros(3)
        # logistic regression with sgd
        (n_epochs, E_out) = \
            logistic_regression(X, line, y, w_init, 2000, 0.01, 0.01, 10000)
        n_epochs_runs = np.append(n_epochs_runs, n_epochs)
        E_out_runs = np.append(E_out_runs, E_out)
    n_epochs_mean = np.mean(n_epochs_runs)
    E_out_mean = np.mean(E_out_runs)
    print(f"Mean epochs = {n_epochs_mean} , mean E_out = {E_out_mean}")
    time_end = time.time()
    print(f"Run time = {time_end - time_start} seconds")

In [9]:
run_exp(100, 2, 100)

Mean epochs = 344.61 , mean E_out = 0.10423824951680664
Run time = 85.0728600025177 seconds


Answer is 0.100 i.e. [d].

---

<img src='hw5_images/hw5p9.png'>

**Answer:**
    
Answer is 344.61 i.e. [a].

---

<img src='hw5_images/hw5p10.png'>

**Answer:**

For PLA, $h(x) = \text{sign}(w^T x)$

PLA update: $\Delta_w = y_n x_n$

We do an update only if a point is misclassified. Meaning the signs of $y_n$ and $w^Tx$ differ, i.e. $y_n w^Tx_n < 0$.

No update (no change of weights) if the point is not misclassified. This means the error is 0.

Thus a good error function is $\max(0, -y_n w^Tx_n) = -\min(0, -y_n w^Tx_n)$
 
$e_n(w) = -\min(0, y_n w^Tx_n)$

Answer is [e].

---